Added tab cycling to Completer

cameron-toy opened the pull request:
![cycle2](https://user-images.githubusercontent.com/38936057/115947690-2c3d9500-a47e-11eb-8eb9-c76672686715.gif)

Added tab and shift + tab cycling to auto completion.

## References

[Cycle completion with tab and automatically insert parenthesis #9082](https://github.com/jupyterlab/jupyterlab/issues/9082)
[Autocomplete cycle options using <KEY> #5863](https://github.com/jupyterlab/jupyterlab/issues/5863)

## User-facing changes
Cycling through auto-complete entries with the arrow keys or the tab key wraps bottom->top and top->bottom. Not sure if this should be the behavior for arrow keys and tab cycling, just tab cycling, or neither.

Note: I had a one-time issue with the shift-tab key combination opening a tooltip in the console but haven't been able to replicate it.

blink1073 commented:
cc @afshin @krassowski

krassowski commented on packages/completer/src/widget.ts:
```diff
@@ -439,6 +439,8 @@ export class Completer extends Widget {
           this.update();
         }
+
+        this._cycle(event.shiftKey ? 'up' : 'down');
```
I am not sure if having discordant keybindings is a good thing, this is:
- <kbd>shift</kbd> + <kbd>Tab</kbd> means "go up", but
- <kbd>Tab</kbd> meaning "insert active completion item"
might end up confusing users and lead to frustration ("it works one way, why doesn't it work the other way too!").

Would it be possibly better to keep only arrows for cycling for this PR, and in another PR add a capability to change the keybindings from "arrows for cycling and Tab for insert" to "Tab/Shift Tab for cycling and Enter for insert"?

krassowski replied:
Nevermind I misunderstood it.

krassowski reviewed (approved):
Looks great to me.

krassowski reviewed (changes requested):
While the UX and code looks great, the tests need to be updated too. In particular these tests of completer/widget (in test/widget.spec.ts) are failing because the behaviour changed:

```
✕ should select the item below and not progress past last
✕ should select the completion item below and not progress past last
✕ should select the item above and not progress beyond first
✕ should select the completion item above and not progress beyond first
```

cameron-toy commented:
> While the UX and code looks great, the tests need to be updated too. In particular these tests of completer/widget (in test/widget.spec.ts) are failing because the behaviour changed:
> 
> ```
> ✕ should select the item below and not progress past last
> ✕ should select the completion item below and not progress past last
> ✕ should select the item above and not progress beyond first
> ✕ should select the completion item above and not progress beyond first
> ```

Okay, updated tests. I also added tests specifically for tab cycling.

Also not too sure about the names.

krassowski commented:
> Also not too sure about the names.

"wrap to" sounds fine as a reference to [wraparound](https://en.wikipedia.org/wiki/Wraparound_(video_games)); alternatively "cycle to" would be fine too (for consistency with the documentation of `_cycle` function).

blink1073 reviewed (approved):
Thanks!
